Client/Frontend/UserContent/UserScripts/RequestBlocking.js

```
@@ -0,0 +1,100 @@
+    }
+    catch(err) {
+      //[DEBUG]console.log(`[RB] URL Error: '${err}'`)
+      return
+    }
+  }
+  
+  const { fetch: originalFetch } = window
+  window.fetch = function() {
+    const [resource, config] = arguments
+    
+    // Extract the url string
+    let urlString
+    if (typeof resource === 'string' || resource instanceof String) {
+      // Simple send where resource is a string
+      urlString = resource
+    } else if (resource !== undefined && (typeof resource.url === 'string' || resource.url instanceof String)) {
+      // Assume its a Request
+      urlString = resource.url
+    } else {
+      //[DEBUG]console.log('[RB] Cannot extract url')
+      return originalFetch.apply(this, arguments)
+    }
+
+    // Resolve the url in case it's a relative path
+    const url = makeURLFromString(urlString)
+    if (url === undefined) { return }
+    //[DEBUG]console.log(`[RB] URL: '${url}'`)
+    
+    return sendMessage(url, 'script').then(blocked => {
+      if (blocked) {
+        //[DEBUG]console.log(`[RB] Request BLOCKED for '${url}'`)
+        return new Promise(function(_, reject) {
+          reject(Error(`Content blocked for '${url}'`))
+        });
+      } else {
+        return originalFetch.apply(this, arguments)
+      }
+    })
+  }
+  
+  const originalOpen = XMLHttpRequest.prototype.open
+  XMLHttpRequest.prototype.open = function() {
+    const sendArguments = arguments
+    originalOpen.apply(this, arguments)
+
+    // Extract the url string
+    let urlString = arguments[1]
+    if (!(typeof urlString === 'string' || urlString instanceof String)) {
+      return
+    }
+
+    let isAsync = arguments[2]
+    if (isAsync !== undefined && !isAsync) {
+      // We can't handle non-async requests as they are not handled via callbacks
+      return
+    }
+
+    // Resolve the url in case it's a relative path
+    const url = makeURLFromString(urlString)
+    if (url === undefined) { return }
```

why return `undefined`?  This seems like invalid input.  `throw`ing seems right here, instead of silent failure